### PR TITLE
TESTS: Check for Netty resource leaks

### LINKED ISSUE #20398: Enable Netty 4 advanced leak detection in tests and fail the build if there is a leak
Netty has pooled and unpooled byte buf alloactors. The pooled allocator uses reference counting to track when a byte buf can be returned to the pool, and leak detection to ensure that byte bufs are properly accessed/released. Netty has different levels of detection from disabled to paranoid (check every access). We should enable leak detection in tests and fail the build if a leak is detected.

Relates #19526, relates #19562

### PR DESCRIPTION
**Pull request by original-brownbear**
* Enabled advanced leak detection when loading `EsTestCase`
* Added custom `Appender` to collect leak logs and check for logged errors in a way similar to what is done for the `StatusLogger`
* Fixes #20398

---------

I manually verified this works by adding a `.retain()` call in `org.elasticsearch.transport.netty4.Netty4MessageChannelHandler#channelRead`.

### CODE REVIEW

**Comment by original-brownbear** [comment]
@jasontedor can you take a look here and let me know if this is what you were looking for in #20398 ? :)

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 1]
```diff
@@ -188,6 +194,25 @@ public static void resetPortCounter() {
...
+            public void append(LogEvent event) {
+                String message = event.getMessage().getFormattedMessage();
+                if (Level.ERROR.equals(event.getLevel()) && message.contains("LEAK:")) {
+                    synchronized (nettyLoggedLeaks) {
+                        nettyLoggedLeaks.add(message);
+                    }
+                }
+            }
+        };
+        leakAppender.start();
```
Comment (jasontedor) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Let's stop this on shutdown.

**Review by jasontedor — APPROVED** [review]
The approach is great! I left one comment, but this LGTM and doesn't need another review.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java (hunk 1 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: done in https://github.com/elastic/elasticsearch/pull/31861/commits/43ed1d8426c8d19b2ec292eca2feaf40d1f680f6

**Comment by original-brownbear** [comment]
@jasontedor thanks!

**Pull request merged**